### PR TITLE
Bamboo scopes and remote api

### PR DESCRIPTION
**Pull request by mappjzc**
### Summary
`curl --location --request GET 'devlake:8080/plugins/bamboo/connections/3/remote-scopes?groupId=&pageToken='`

```json
{
	"children": [{
		"type": "scope",
		"parentId": null,
		"id": "TEST1",
		"name": "test_project",
		"data": {
			"connectionId": 0,
			"projectKey": "TEST1",
			"name": "test_project",
			"description": "just a test project",
			"link": "http://54.172.92.89:8085/rest/api/latest/project/TEST1",
			"rel": ""
		}
	}],
	"nextPageToken": ""
}
```

`curl --location --request GET 'devlake:8080/plugins/bamboo/connections/3/search-remote-scopes?search=test&page=1&pageSize=100'`

```json
{
	"children": [{
		"type": "scope",
		"parentId": null,
		"id": "",
		"name": "",
		"data": {
			"connectionId": 0,
			"projectKey": "",
			"name": "",
			"description": "",
			"link": "http://54.172.92.89:8085/rest/api/latest/project/",
			"rel": ""
		}
	}],
	"page": 1,
	"pageSize": 100
}
```

### Does this close any open issues?
related #3322

### Screenshots
image
image
image
.
image
image

image
image
image
image

### CODE REVIEW

File: backend/plugins/bamboo/api/blueprint_V200_test.go [hunk 1]
```diff
@@ -94,13 +95,17 @@ func TestMakeDataSourcePipelinePlanV200(t *testing.T) {
...
 
 	var expectPlans = plugin.PipelinePlan{
 		{
 			{
-				Plugin:   "bamboo",
-				Subtasks: []string{},
+				Plugin: "bamboo",
+				Subtasks: []string{
+					tasks.ConvertProjectsMeta.Name,
+				},
```
Comment (warren830) [inline backend/plugins/bamboo/api/blueprint_V200_test.go:108]: Why do you change this?

File: backend/plugins/bamboo/api/blueprint_v200.go [hunk 2]
```diff
@@ -57,6 +57,11 @@ func MakePipelinePlanV200(
 		return nil, nil, err
 	}
 
+	// 测试
+	fmt.Printf("MakePipelinePlanV200\r\n")
+	fmt.Printf("pp:%+v\r\n", pp)
+	fmt.Printf("sc:%+v\r\n", sc)
```
Comment (warren830) [inline backend/plugins/bamboo/api/blueprint_v200.go:63]: Please delete this block

File: backend/plugins/bamboo/api/remote.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	Page     int                 `json:"page"`
+	PageSize int                 `json:"pageSize"`
+}
+
+type PageData struct {
+	Page     int `json:"page"`
+	PageSize int `json:"per_page"`
+}
+
+type GroupResponse struct {
```
Comment (warren830) [inline backend/plugins/bamboo/api/remote.go:59]: Bamboo doesn’t have group

File: backend/plugins/bamboo/api/scope.go [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+	keeper := make(map[string]struct{})
+	now := time.Now()
+	for _, project := range projects.Data {
+		if _, ok := keeper[project.ProjectKey]; ok {
+			return nil, errors.BadInput.New("duplicated item")
+		} else {
+			keeper[project.ProjectKey] = struct{}{}
+		}
+		project.ConnectionId = connectionId
+		project.CreatedAt = now
```
Comment (warren830) [inline backend/plugins/bamboo/api/scope.go:74]: CreatedAt will be generated by db

File: backend/plugins/bamboo/impl/impl.go [hunk 5]
```diff
@@ -63,6 +63,8 @@ func (p Bamboo) TransformationRule() interface{} {
 }
 
 func (p Bamboo) MakeDataSourcePipelinePlanV200(connectionId uint64, scopes []*plugin.BlueprintScopeV200, syncPolicy plugin.BlueprintSyncPolicy) (plugin.PipelinePlan, []plugin.Scope, errors.Error) {
+	// 测试
+	fmt.Printf("MakeDataSourcePipelinePlanV200\r\n")
```
Comment (warren830) [inline backend/plugins/bamboo/impl/impl.go:67]: Please delete this

File: backend/plugins/bamboo/impl/impl.go [hunk 6]
```diff
@@ -186,11 +188,27 @@ func (p Bamboo) ApiResources() map[string]map[string]plugin.ApiResourceHandler {
...
+			"GET":   api.GetScope,
+			"PATCH": api.UpdateScope,
+		},
 	}
 }
 
 func (p Bamboo) MakePipelinePlan(connectionId uint64, scope []*plugin.BlueprintScopeV100) (plugin.PipelinePlan, errors.Error) {
-	return nil, errors.Default.New("Bamboo does not support blueprint v100")
+	// 测试
+	fmt.Printf("MakePipelinePlan\r\n")
```
Comment (warren830) [inline backend/plugins/bamboo/impl/impl.go:210]: Please delete

File: backend/plugins/bamboo/models/job.go [hunk 7]
```diff
@@ -37,17 +37,16 @@ type BambooJob struct {
...
-		JobKey:      apiRes.SearchEntity.Key,
-		Id:          apiRes.Id,
-		Name:        apiRes.SearchEntity.JobName,
-		PlanName:    apiRes.SearchEntity.PlanName,
-		ProjectName: apiRes.SearchEntity.ProjectName,
-		Description: apiRes.SearchEntity.Description,
-		BranchName:  apiRes.SearchEntity.BranchName,
-		Type:        apiRes.SearchEntity.Type,
-	}
+func (b *BambooJob) Convert(apiRes *ApiBambooJob) *BambooJob {
```
Comment (warren830) [inline backend/plugins/bamboo/models/job.go:40]: Why do you change this?

File: backend/plugins/bamboo/models/plan.go [hunk 8]
```diff
@@ -43,26 +43,25 @@ type BambooPlan struct {
...
-		ShortKey:                  apiProject.ShortKey,
-		Type:                      apiProject.Type,
-		Enabled:                   apiProject.Enabled,
-		Href:                      apiProject.Href,
-		Rel:                       apiProject.Rel,
-		IsFavourite:               apiProject.IsFavourite,
-		IsActive:                  apiProject.IsActive,
-		IsBuilding:                apiProject.IsBuilding,
-		AverageBuildTimeInSeconds: apiProject.AverageBuildTimeInSeconds,
-	}
```
Comment (warren830) [inline backend/plugins/bamboo/models/plan.go:65]: Why do Hou change this?

File: backend/plugins/bamboo/tasks/job_extractor.go [hunk 9]
```diff
@@ -44,7 +45,7 @@ func ExtractJob(taskCtx plugin.SubTaskContext) errors.Error {
 			if err != nil {
 				return nil, err
 			}
-			body := models.BambooJob{}.Convert(res)
+			body := (&models.BambooJob{}).Convert(res)
```
Comment (warren830) [inline backend/plugins/bamboo/tasks/job_extractor.go:48]: Why do you change this?

File: backend/plugins/bamboo/api/blueprint_V200_test.go (hunk 1 above)
Comment (mappjzc) [reply backend/plugins/bamboo/api/blueprint_V200_test.go:108]: we need subtask to unit test the subtask create part.

File: backend/plugins/bamboo/models/job.go (hunk 7 above)
Comment (mappjzc) [reply backend/plugins/bamboo/models/job.go:40]: a := &BambooJob{}
b := a.Convert()

that will make a and b to be diffrent.
So it need to be fixed.

File: backend/plugins/bamboo/models/plan.go (hunk 8 above)
Comment (mappjzc) [reply backend/plugins/bamboo/models/plan.go:65]: a := &BambooPlan{}
b := a.Convert()

that will make a and b to be diffrent.
So it need to be fixed

File: backend/plugins/bamboo/tasks/job_extractor.go (hunk 9 above)
Comment (mappjzc) [reply backend/plugins/bamboo/tasks/job_extractor.go:48]: for Convert changed

File: backend/plugins/bamboo/api/remote.go [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+		groupId = []string{""}
+	}
+
+	pageToken, ok := input.Query["pageToken"]
+	if !ok || len(pageToken) == 0 {
+		pageToken = []string{""}
+	}
+
+	// get gid and pageData
+	gid := groupId[0]
```
Comment (warren830) [inline backend/plugins/bamboo/api/remote.go:98]: As bamboo doesn't have group, I think we don't need gid

File: backend/plugins/bamboo/api/remote.go [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+	}
+
+	return pt, nil
+}
+
+func GetQueryFromPageData(pageData *PageData) (url.Values, errors.Error) {
+	query := url.Values{}
+	query.Set("showEmpty", fmt.Sprintf("%v", true))
+	query.Set("max-result", fmt.Sprintf("%v", pageData.PageSize))
+	query.Set("start-index", fmt.Sprintf("%v", (pageData.Page-1)*pageData.PageSize))
```
Comment (warren830) [inline backend/plugins/bamboo/api/remote.go:307]: use pageData.skip to replace (pageData.Page-1)*pageData.PageSize

File: backend/plugins/bamboo/api/remote.go [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+	pt := &PageData{}
+	err = json.Unmarshal(pageTokenDecode, pt)
+	if err != nil {
+		return nil, errors.Default.Wrap(err, fmt.Sprintf("json Unmarshal pageTokenDecode failed %s", pageTokenDecode))
+	}
+
+	return pt, nil
+}
+
+func GetQueryFromPageData(pageData *PageData) (url.Values, errors.Error) {
```
Comment (warren830) [inline backend/plugins/bamboo/api/remote.go:303]: url.Values is enough

File: backend/plugins/bamboo/impl/impl.go [hunk 13]
```diff
@@ -186,11 +187,25 @@ func (p Bamboo) ApiResources() map[string]map[string]plugin.ApiResourceHandler {
...
+		"connections/:connectionId/scopes/:projectKey": {
+			"GET":   api.GetScope,
+			"PATCH": api.UpdateScope,
+		},
 	}
 }
 
 func (p Bamboo) MakePipelinePlan(connectionId uint64, scope []*plugin.BlueprintScopeV100) (plugin.PipelinePlan, errors.Error) {
-	return nil, errors.Default.New("Bamboo does not support blueprint v100")
+	return nil, errors.Default.New("Bamboo don't support blueprint v100")
```
Comment (warren830) [inline backend/plugins/bamboo/impl/impl.go:208]: `does not` is right

File: backend/plugins/bamboo/models/job.go (hunk 7 above)
Comment (warren830) [reply backend/plugins/bamboo/models/job.go:40]: I don't think any one would do that thing, well, if you change it to this, BambooJob will have two methods, one's receiver is pointer, one's is value. Hmm, maybe we'd better use func ConvertJob() is better

Comment (warren830) [reply backend/plugins/bamboo/models/job.go:40]: If you want to keep this, please delete return value

**Review by warren830 — APPROVED** [review]
LGTM

**Pull request merged**